From 47b4778a07bd40f707493f56456e24af63f00c73 Mon Sep 17 00:00:00 2001
From: Sagar Cheluvegowda <quic_scheluve@quicinc.com>
Date: Mon, 24 Mar 2025 13:48:54 -0700
Subject: [PATCH 07/16] PENDING: net: stmmac: Fix USXGMII link up sequence

PCS Rx link is failing intermittently on sa8797p,
Fix USXGMII link up sequence to address this issue.

Change-Id: Iaf922b1ee0d3f8bc4a3ae6e96190ad66c9f1c5ec
Signed-off-by: Sagar Cheluvegowda <quic_scheluve@quicinc.com>
(cherry picked from commit f7a0b5a006b29cb8ac838e341c539ad69b6c42b1)
---
 drivers/net/pcs/pcs-xpcs-qcom.c | 40 +++++++++++++++++++++++++++++++--
 include/linux/pcs-xpcs-qcom.h   |  1 +
 2 files changed, 39 insertions(+), 2 deletions(-)

diff --git a/drivers/net/pcs/pcs-xpcs-qcom.c b/drivers/net/pcs/pcs-xpcs-qcom.c
index 4c213f44eac4..a004d74a88a5 100644
--- a/drivers/net/pcs/pcs-xpcs-qcom.c
+++ b/drivers/net/pcs/pcs-xpcs-qcom.c
@@ -367,7 +367,7 @@ static int xpcs_config_aneg_c37(struct dw_xpcs_qcom *qxpcs)
 	if (ret < 0)
 		return -EINVAL;
 
-	ret |= DW_VR_MII_TX_CONFIG_MASK;
+	ret |= DW_VR_MII_TX_CONFIG_PHY_SIDE << DW_VR_MII_AN_CTRL_TX_CONFIG_SHIFT;
 	ret |= DW_VR_MII_SGMII_LINK_STS;
 
 	qcom_xpcs_write(qxpcs, DW_VR_MII_AN_CTRL, ret);
@@ -386,6 +386,7 @@ static int qcom_xpcs_do_config(struct dw_xpcs_qcom *qxpcs, phy_interface_t inter
 
 	switch (compat->an_mode) {
 	case DW_AN_C37_USXGMII:
+		qxpcs->needs_aneg = true;
 		ret = xpcs_config_aneg_c37(qxpcs);
 		if (ret < 0)
 			return ret;
@@ -451,10 +452,35 @@ static void qcom_xpcs_get_state(struct phylink_pcs *pcs,
 void qcom_xpcs_link_up_usxgmii(struct dw_xpcs_qcom *qxpcs, int speed)
 {
 	int mmd_ctrl;
+	int ret;
+
+	ret = qcom_xpcs_read(qxpcs, DW_VR_MII_PCS_DIG_CTRL1);
+	if (ret < 0)
+		goto read_err;
+
+	qcom_xpcs_write(qxpcs, DW_VR_MII_PCS_DIG_CTRL1, ret | SOFT_RST);
+
+	mmd_ctrl = qcom_xpcs_poll_reset(qxpcs, DW_VR_MII_PCS_DIG_CTRL1,
+					SW_RST_BIT_STATUS);
+	if (mmd_ctrl < 0) {
+		XPCSERR("Failed to perform soft reset\n");
+		return;
+	}
+
+	if (qxpcs->needs_aneg) {
+		ret = qcom_xpcs_read(qxpcs, DW_VR_MII_AN_CTRL);
+		if (ret < 0)
+			goto read_err;
+
+		ret |= DW_VR_MII_TX_CONFIG_PHY_SIDE << DW_VR_MII_AN_CTRL_TX_CONFIG_SHIFT;
+		ret |= DW_VR_MII_SGMII_LINK_STS;
+
+		qcom_xpcs_write(qxpcs, DW_VR_MII_AN_CTRL, ret);
+	}
 
 	mmd_ctrl = qcom_xpcs_read(qxpcs, DW_SR_MII_MMD_CTRL);
 	if (mmd_ctrl < 0)
-		goto out;
+		goto read_err;
 
 	mmd_ctrl &= ~DW_USXGMII_SS_MASK;
 
@@ -489,12 +515,22 @@ void qcom_xpcs_link_up_usxgmii(struct dw_xpcs_qcom *qxpcs, int speed)
 
 	qcom_xpcs_write(qxpcs, DW_SR_MII_MMD_CTRL, mmd_ctrl);
 
+	if (qxpcs->needs_aneg) {
+		ret = qcom_xpcs_read(qxpcs, DW_SR_MII_MMD_CTRL);
+		if (ret < 0)
+			goto read_err;
+
+		qcom_xpcs_write(qxpcs, DW_SR_MII_MMD_CTRL, ret | AN_CL37_EN);
+	}
+
 	mmd_ctrl = qcom_xpcs_reset_usxgmii(qxpcs);
 	if (mmd_ctrl < 0)
 		goto out;
 
 	XPCSINFO("USXGMII link is up\n");
 	return;
+read_err:
+	XPCSERR("Failed to read register\n");
 out:
 	XPCSERR("Failed to bring up USXGMII link\n");
 }
diff --git a/include/linux/pcs-xpcs-qcom.h b/include/linux/pcs-xpcs-qcom.h
index 2bdb9b65f85c..5d526a04780c 100644
--- a/include/linux/pcs-xpcs-qcom.h
+++ b/include/linux/pcs-xpcs-qcom.h
@@ -23,6 +23,7 @@ struct dw_xpcs_qcom {
 	void __iomem *addr;
 	int pcs_intr;
 	bool intr_en;
+	bool needs_aneg;
 	int phy_interface;
 };
 
